File: current_session/c++/213.cpp
// clang++ -std=c++17 c++/213.cpp -o 213 && ./213 -v
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int rob(vector<int>& nums) {
    // use DP
    // remove the minimum house then treat it as linear one
    if (nums.size() == 1) {return nums[0];}
    int L = nums.size();
    return max(finder(0, L-1, nums), finder(1, L, nums));
  }

  // search from start to end-1
  int finder(int start, int end, vector<int>& nums) {
    int rob = 0, notRob = 0;
    for (int j = start; j < end; ++j) {
      int tmp = notRob + nums[j];
      notRob = max(notRob, rob);
      rob = tmp;
    }
    return max(rob, notRob);
  }
};

int main() {
  vector<int> test1 = {2,3,2};
  vector<int> test2 = {1,7,9,2};
  vector<int> test3 = {1,2,3,1};
  cout << Solution().rob(test1) << endl;
  cout << Solution().rob(test2) << endl;
  cout << Solution().rob(test3) << endl;
}

File: current_session/java/213.java
/*
cp 213.java Solution.java | javac Solution.java | java Solution
*/
import java.util.Arrays;

class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        int[] a = Arrays.copyOfRange(nums, 0, nums.length-1);
        int[] b = Arrays.copyOfRange(nums, 1, nums.length);
        return Math.max(finder(a), finder(b));
    }
    private int finder(int[] houses) {
        int rob = 0, notRob = 0;
        for (int i = 0; i < houses.length; ++i ) {
            int tmp = notRob + houses[i];
            notRob = Math.max(rob, notRob);
            rob = tmp;
        }
        return Math.max(rob, notRob);
    }

    public static void main(String[] args) {
        int[] input = {2,3,2};
        Solution sol = new Solution();
        System.out.println(sol.rob(input));
    }
}